source/Symbol/SwiftASTContext.cpp

```diff
@@ -987,13 +989,29 @@ static bool DeserializeAllCompilerFlags(SwiftASTContext &swift_ast,
                         ast_file_data_sp->GetByteSize());
     while (!buf.empty()) {
       std::string last_sdk_path;
-      auto info = swift::serialization::validateSerializedAST(buf);
-      if ((info.status != swift::serialization::Status::Valid) ||
-          (info.bytes == 0) || (info.bytes > buf.size())) {
-        if (log)
-          log->Printf("Unable to load AST for module %s from library: %s.",
-                      info.name.str().c_str(),
-                      module.GetSpecificationDescription().c_str());
+      swift::serialization::ExtendedValidationInfo extended_validation_info;
+      swift::serialization::ValidationInfo info =
+          swift::serialization::validateSerializedAST(
+              buf, &extended_validation_info);
+      bool InvalidAST = info.status != swift::serialization::Status::Valid;
+      bool InvalidSize = (info.bytes == 0) || (info.bytes > buf.size());
+      if (InvalidAST) {
+        swift::ASTContext &ast_ctx = *swift_ast.GetASTContext();
+        StringRef module_spec = module.GetSpecificationDescription();
+        swift::Identifier module_id = ast_ctx.getIdentifier(module_spec);
+        std::unique_ptr<swift::ModuleFile> loaded_module_file;
+        std::unique_ptr<llvm::MemoryBuffer> module_input_buffer =
+            llvm::MemoryBuffer::getMemBuffer(buf, module_spec,
+                                             /*NullTerminator=*/false);
+        swift::ModuleFile::load(std::move(module_input_buffer), {}, false,
+                                loaded_module_file);
+        swift::serialization::diagnoseSerializedASTLoadFailure(
+            ast_ctx, swift::SourceLoc(), info, extended_validation_info,
+            module_spec, "<invalid-doc-id>", loaded_module_file.get(),
+            module_id);
```

Comment:
Actually, now that you've made this take an unowned module, you could support passing `NULL` here. We only use it for missing dependencies, which aren't checked by `validateSerializedAST` (because that would go do more work).

Reply:
If there's a need to call diagnoseSerializedASTLoadFailure with a null ModuleFile, I think it'd be cleaner to address it in a follow-up. I don't see one at the moment (I believe ModuleFile::load always allocates), so I'd prefer to keep the swift change as nfc code movement.

Reply:
Sure, but `ModuleFile::load` isn't cheap, so it's probably a worthwhile follow-up.